From 1b7f7f0b835662c062282eb4ba2616944cb0ac69 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Wed, 1 Oct 2014 17:40:32 -0400
Subject: [ethernet] cxgb4: Add support to S25FL032P flash

Message-id: <1412185238-11567-2-git-send-email-svemuri@redhat.com>
Patchwork-id: 96827
O-Subject: [RHEL7.1 PATCH BZ 1147683 1/7] cxgb4: Add support to S25FL032P flash
Bugzilla: 1147683
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>
RH-Acked-by: David S. Miller <davem@redhat.com>

Add support for Spansion S25FL032P flash
Based on original work by Dimitris Michailidis

Upstream : https://git.kernel.org/cgit/linux/kernel/git/davem/net-next.git/commit/drivers/net/ethernet/chelsio?id=fe2ee139aa295936e2bf6b7c620624d871c65058

Signed-off-by: Hariprasad Shenai <hariprasad@chelsio.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit fe2ee139aa295936e2bf6b7c620624d871c65058)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c b/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c
index 85dfe73..451c2d9 100644
--- a/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c
+++ b/drivers/net/ethernet/chelsio/cxgb4/t4_hw.c
@@ -3791,8 +3791,20 @@ int t4_wait_dev_ready(struct adapter *adap)
  return t4_read_reg(adap, PL_WHOAMI) != 0xffffffff ? 0 : -EIO;
 }
 
+struct flash_desc {
+ u32 vendor_and_model_id;
+ u32 size_mb;
+};
+
 static int get_flash_params(struct adapter *adap)
 {
+ /* Table for non-Numonix supported flash parts.  Numonix parts are left
+  * to the preexisting code.  All flash parts have 64KB sectors.
+  */
+ static struct flash_desc supported_flash[] = {
+  { 0x150201, 4 << 20 },       /* Spansion 4MB S25FL032P */
+ };
+
  int ret;
  u32 info;
 
@@ -3803,6 +3815,14 @@ static int get_flash_params(struct adapter *adap)
  if (ret)
   return ret;
 
+ for (ret = 0; ret < ARRAY_SIZE(supported_flash); ++ret)
+  if (supported_flash[ret].vendor_and_model_id == info) {
+   adap->params.sf_size = supported_flash[ret].size_mb;
+   adap->params.sf_nsec =
+    adap->params.sf_size / SF_SEC_SIZE;
+   return 0;
+  }
+
  if ((info & 0xff) != 0x20)             /* not a Numonix flash */
   return -EINVAL;
  info >>= 16;                           /* log2 of size */
-- 
1.7.1